### PR TITLE
feat: Deterministic builds with receipt cache (+misc)

### PR DESCRIPTION
**Pull request by hashcashier**
This PR improves the usability of zeth as follows:
* [updates the README](https://github.com/risc0/zeth/blob/a41975135ca59ce81025c67204f23f9758ecd117/README.md) to use the new CLI usage.
* Caches receipts using (image id, input hash, output hash) as identifiers for faster re-proving.
* Improves logging to reduce `info`-level spam (messages moved to debug)
* Enable deterministic builds by default (disable with `debug-guest-build` feature)

### CODE REVIEW

File: lib/src/builder/execute/optimism.rs [hunk 1]
```diff
@@ -69,19 +69,18 @@ impl TxExecStrategy<OptimismTxEssence> for OpTxExecStrategy {
...
-            info!("  Base fee per gas: {}", header.base_fee_per_gas);
-            info!("  Extra data: {:?}", block_builder.input.extra_data);
+            trace!("Block no. {}", header.number);
+            trace!("  EVM spec ID: {:?}", spec_id);
+            trace!("  Timestamp: {}", dt);
+            trace!("  Transactions: {}", block_builder.input.transactions.len());
+            trace!("  Fee Recipient: {:?}", block_builder.input.beneficiary);
+            trace!("  Gas limit: {}", block_builder.input.gas_limit);
+            trace!("  Base fee per gas: {}", header.base_fee_per_gas);
+            trace!("  Extra data: {:?}", block_builder.input.extra_data);
```
Comment (Wollac) [inline lib/src/builder/execute/optimism.rs:83]: This is much less verbose, so I think it should be debug

File: primitives/src/transactions/ethereum.rs [hunk 2]
```diff
@@ -334,11 +334,12 @@ impl Encodable for TransactionKind {
...
-            buf.advance(1);
-            Ok(TransactionKind::Create)
-        } else {
-            Ok(TransactionKind::Call(Address::decode(buf)?))
+        match buf.first().copied() {
+            Some(EMPTY_STRING_CODE) => {
+                buf.advance(1);
+                Ok(TransactionKind::Create)
+            }
+            _ => Ok(TransactionKind::Call(Address::decode(buf)?)),
```
Comment (Wollac) [inline primitives/src/transactions/ethereum.rs:342]:
```suggestion
            _ => Address::decode(buf).map(TransactionKind::Call),
```

**Review by Wollac — APPROVED** [review]

File: lib/src/builder/execute/optimism.rs (hunk 1 above)
Comment (hashcashier) [reply lib/src/builder/execute/optimism.rs:83]: it gets really spammy and repetitive, especially during composition, which already shows block number and hash

Comment (Wollac) [reply lib/src/builder/execute/optimism.rs]: That's why I like it being debug and not info, it only comes once per block, the logs on tx level are even more frequent and trace

**Pull request merged**